Backend/FcEvaluate.hs

```diff
@@ -142,25 +144,24 @@ groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
 groundTerm (FcTmTyAbs ty t1) = FcTmTyApp (FcTmTyAbs ty (groundTerm t1)) FcTyUnit
 groundTerm t                 = t
 
--- | Fully evaluates given term using lazy semantics. However if the final
--- | result is a term application, then the right term in the application may
--- | not be fully evaluated. In order to produce a clear, readable output we
--- | choose to evaluate all branches recursively before returning the result.
+-- | Fully evaluates given term. If final result is a term application, then
+-- | the right hand side term in the application may not be fully evaluated.
+-- | In order to produce a clear, readable output we choose to evaluate all
+-- | branches recursively before returning the result.
 fullEval :: FcTerm 'Fc -> EvM (Either String (FcTerm 'Fc))
 fullEval t = fullStep t >>= \case
   Right (FcTmApp t1 t2) -> fullEval t2 >>= \case
     Right t2' -> fullStep (FcTmApp t1 t2')
     Left err  -> return $ Left err
   result -> return $ result
```

Comment:
Actually, I think this function is wrong. A data constructor can have multiple arguments. We should start evaluating them from left to right. You should try and recursively evaluate `t1` first, and then `t2`.

Reply:
I don't think that is necessary because the congruence rule will evaluate `t1` till it goes no further. Once `t1` is in normal form we receive the `FcTmApp` here.

Reply:
Added a note